### ISSUE TITLE
cli+ui: create safety rails for deployment

### ISSUE THREAD

**Issue by dianasaur323** [cockroachdb/cockroach#12157]
For the purpose of supporting "safe practices" when deploying CockroachDB, we would like to incorporate greater visibility into deployment configurations as well as improved UX in the CLI. An example would be more clearly surfacing configurations in the admin UI, or improving the UX of assigning certificates on install. We basically want to provide a general sense of security for our users so that they can configure their clusters and check that the configurations are set as expected. Comments and specific feature ideas would be appreciated. For context, this is to help flesh out what items should make it onto Q1 development, if any.

**Comment by knz** [comment]
Let's expand the items in a list:

- [ ] show the node configuration, including command-line flags and environment variables, in the admin UI, and perhaps via the CLI too
- [ ] allow the user to generate a detailed configuration report automatically, which they can use and report to us when troubleshooting
- [ ] report in the admin UI how well the certificates have been set up ("security checkmark")
- [ ] provide a step-by-step procedure/workflow in the CLI and/or admin UI to set up certificates

**Comment by knz** [comment]
@mberhault @bdarnell do you have suggestions as to how to expand Diana's idea to "improving the UX of assigning certificates on install" to actionable enhancement requests?

Or perhaps 1) do you know who has the best opinions on this topic and 2) what other software package have a (remotely) similar security story to ours, that we can look at for inspiration?

**Comment by dianasaur323** [comment]
@knz regarding the UX of assigning certificates on install, Peter mentioned Docker as an example. I don't have specifics yet as that isn't something I've dug into, but hopefully that helps a little.

**Comment by petermattis** [comment]
https://blog.docker.com/2016/06/docker-1-12-built-in-orchestration/:

> Security
>
>A core principle for Docker 1.12 is creating a zero configuration, secure-by-default, out of the box experience for the Docker platform. One of the major hurdles that administrators often face with deploying applications into production is running them securely, Docker 1.12 allows an administrator to follow the exact same steps setting up a demo cluster that they would to setup a secure production cluster.
>
> Security is not something you can bolt-on after the fact. That is why Docker 1.12 comes with mutually authenticated TLS, providing authentication, authorization and encryption to the communications of every node participating in the swarm, out of the box.
> 
> When starting your first manager, Docker Engine will generate a new Certificate Authority (CA) and a set of initial certificates for you. After this initial step, every node joining the swarm will automatically be issued a new certificate with a randomly generated ID, and their current role in the swarm (manager or worker). These certificates will be used as their cryptographically secure node identity for the lifetime of their participation in this swarm, and will be used by the managers to ensure secure dissemination of tasks and other updates.
> 
> One of the biggest barriers of adoption of TLS has always been the difficulty of creating, configuring and maintaining the necessary Public Key Infrastructure (PKI). With Docker 1.12, everything not only gets setup and configured with safe defaults for you, but we also automated one of the most painful parts of dealing with TLS certificates: certificate rotation.
>
> Under the hood, every node participating in the swarm is constantly refreshing its certificates, ensuring that potentially leaked or compromised certificates are no longer valid. The frequency with which certificates are rotated can be configured by the user, and set as low as every 30 minutes.
>
> If you would like to use your own Certificate Authority, we also support an external-CA mode, where the managers in the swarm simply relay the Certificate Signing Requests of the nodes attempting to join the cluster to a remote URL.

**Comment by bdarnell** [comment]
My only firsthand experience with certificate-management systems were all highly custom in-house systems at previous employers. Most of the public systems I've looked at seem to involve a lot of manual work (e.g. [kubernetes](https://coreos.com/kubernetes/docs/latest/openssl.html)). 

I think that ultimately we're going to need some sort of tool that combines creating a VM/container/whatever and its certificate. This is what the `cockroach-prod` repo was for before we switched to terraform. Maybe [terraform can do this for us too?](https://www.terraform.io/docs/providers/tls/).  

The hard part, as always, is key management - where does the CA key live? We should probably support storing it in something like Vault instead of just a plain file. Incidentally, Vault does something nice here: instead of a single key file, it generates a [sharded key](https://www.vaultproject.io/intro/getting-started/deploy.html) (so that e.g. 3 of the 5 key shards must be entered to unlock the server at boot).
